### PR TITLE
Issue #12078: Github action Release prepare

### LINKED ISSUE #12078: Github action "Release prepare"
https://github.com/checkstyle/contribution/blob/master/releasenotes-builder/diagrams-manual-trigger.md

we need to make action to do `run Github action "Release Prepare '10.3.1'"` (action takes human provided version as input, just to make sure it can not be executed unintentionally)
if user provider input is not matching pom - fail build
so we simply make script that do "mvn:prepare"
and github action commit and push changes and push created commit by maven.

### PR DESCRIPTION
**Pull request by Rahulkhinchi03**
Fixes #12078 

Final Testing: https://github.com/Rahulkhinchi03/checkstyle-test/actions/runs/2928941890

### CODE REVIEW

File: .ci/maven-prepare.sh [hunk 1]
```diff
@@ -0,0 +1,29 @@
+#!/usr/bin/env bash
+
+set -e
+
+if [[ -z $1 ]]; then
+  echo "latest version is not set"
+  exit 1
+fi
+LATEST_VERSION=$1
+echo LATEST_VERSION="$LATEST_VERSION"
```
Comment (romani) [inline .ci/maven-prepare.sh:10]: Not a latest.
It is target version.

File: .ci/maven-prepare.sh [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+POM_FILE=pom.xml
+CURRENT_VERSION=$(grep "^  <version>.*</version>$" $POM_FILE | awk -F'[><]' '{print $3}')
```
Comment (romani) [inline .ci/maven-prepare.sh:13]: Use xmlstartlet to certain on version from pom.xml.
In pom version is with '-SNAPSHOT'.

File: .ci/maven-prepare.sh [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+echo CURRENT_VERSION="$CURRENT_VERSION"
+
+if [ "$LATEST_VERSION" != "$CURRENT_VERSION" ]; then
+  mvn -e --no-transfer-progress versions:set \
+    -DgroupId=com.puppycrawl.tools -DartifactId=checkstyle -DnewVersion="$LATEST_VERSION-SNAPSHOT"
```
Comment (romani) [inline .ci/maven-prepare.sh:18]: No sets .
On mismatch - fail build.

File: .github/maven_prepare.yml [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+name: "Maven Release Prepare"
+on:
+  workflow_dispatch:
+    inputs:
+      version:
+        description: 'New Version'
+        required: true
+permissions:
+  contents: write
+  pull-requests: write
```
Comment (romani) [inline .github/maven_prepare.yml:14]: Can you remove it from all yml that do nothing with PR creation

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .ci/maven-prepare.sh (hunk 1 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

File: .ci/maven-prepare.sh (hunk 2 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

File: .github/maven_prepare.yml (hunk 4 above)
Comment (Rahulkhinchi03) [reply .github/maven_prepare.yml]: Changes Done

File: .ci/maven-prepare.sh (hunk 2 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

File: .ci/maven-prepare.sh (hunk 3 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

**Comment by Rahulkhinchi03** [comment]
Testing:

```
rahul@rk7:~/Desktop/opensource/checkstyle-test$ ./.ci/maven-prepare.sh 1.1
TARGET_VERSION=1.1
[sudo] password for rahul: 
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
xmlstarlet is already the newest version (1.6.1-2.1).
0 upgraded, 0 newly installed, 0 to remove and 202 not upgraded.
CURRENT_VERSION=1.1
Version bump in pom.xml (release:prepare) ...
[INFO] Error stacktraces are turned on.
[INFO] Scanning for projects...
[INFO] 
[INFO] ----------------< com.puppycrawl.tools:checkstyle-test >----------------
[INFO] Building checkstyle-test 1.1-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-release-plugin:2.5.3:prepare (default-cli) @ checkstyle-test ---
[INFO] Verifying that there are no local modifications...
[INFO]   ignoring changes on: **/pom.xml.releaseBackup, **/pom.xml.next, **/pom.xml.tag, **/pom.xml.branch, **/release.properties, **/pom.xml.backup
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git rev-parse --show-toplevel
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git status --porcelain .
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[WARNING] Ignoring unrecognized line: ?? release.properties
[INFO] Ignoring SNAPSHOT depenedencies and plugins ...
[INFO] Transforming 'checkstyle-test'...
[INFO] Not generating release POMs
[INFO] Executing goals 'package'...
[INFO] [INFO] Scanning for projects...
[INFO] [INFO] 
[INFO] [INFO] ----------------< com.puppycrawl.tools:checkstyle-test >----------------
[INFO] [INFO] Building checkstyle-test 1.1
[INFO] [INFO] --------------------------------[ jar ]---------------------------------
[INFO] [INFO] 
[INFO] [INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ checkstyle-test ---
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] Copying 0 resource
[INFO] [INFO] 
[INFO] [INFO] --- maven-compiler-plugin:3.1:compile (default-compile) @ checkstyle-test ---
[INFO] [INFO] Nothing to compile - all classes are up to date
[INFO] [INFO] 
[INFO] [INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ checkstyle-test ---
[INFO] [INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] [INFO] skip non existing resourceDirectory /home/rahul/Desktop/opensource/checkstyle-test/src/test/resources
[INFO] [INFO] 
[INFO] [INFO] --- maven-compiler-plugin:3.1:testCompile (default-testCompile) @ checkstyle-test ---
[INFO] [INFO] Nothing to compile - all classes are up to date
[INFO] [INFO] 
[INFO] [INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ checkstyle-test ---
[INFO] [INFO] Tests are skipped.
[INFO] [INFO] 
[INFO] [INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ checkstyle-test ---
[INFO] [INFO] Building jar: /home/rahul/Desktop/opensource/checkstyle-test/target/checkstyle-test-1.1.jar
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] BUILD SUCCESS
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] [INFO] Total time:  0.665 s
[INFO] [INFO] Finished at: 2022-08-22T22:10:01+05:30
[INFO] [INFO] ------------------------------------------------------------------------
[INFO] Checking in modified POMs...
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git add -- pom.xml
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git rev-parse --show-toplevel
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git status --porcelain .
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[WARNING] Ignoring unrecognized line: ?? pom.xml.releaseBackup
[WARNING] Ignoring unrecognized line: ?? release.properties
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git commit --verbose -F /tmp/maven-scm-63430675.commit pom.xml
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git symbolic-ref HEAD
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git push git@github.com:Rahulkhinchi03/checkstyle-test.git refs/heads/main:refs/heads/main
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Tagging release with the label 1.1...
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git tag -F /tmp/maven-scm-1208366036.commit 1.1
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git push git@github.com:Rahulkhinchi03/checkstyle-test.git refs/tags/1.1
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git ls-files
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Transforming 'checkstyle-test'...
[INFO] Not removing release POMs
[INFO] Checking in modified POMs...
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git add -- pom.xml
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git rev-parse --show-toplevel
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git status --porcelain .
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[WARNING] Ignoring unrecognized line: ?? pom.xml.releaseBackup
[WARNING] Ignoring unrecognized line: ?? release.properties
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git commit --verbose -F /tmp/maven-scm-1477377805.commit pom.xml
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git symbolic-ref HEAD
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Executing: /bin/sh -c cd /home/rahul/Desktop/opensource/checkstyle-test && git push git@github.com:Rahulkhinchi03/checkstyle-test.git refs/heads/main:refs/heads/main
[INFO] Working directory: /home/rahul/Desktop/opensource/checkstyle-test
[INFO] Release preparation complete.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  11.392 s
[INFO] Finished at: 2022-08-22T22:10:10+05:30
[INFO] ------------------------------------------------------------------------
rahul@rk7:~/Desktop/opensource/checkstyle-test$ 
```

**Comment by Rahulkhinchi03** [comment]
@romani When running it in GItHub Action: https://github.com/Rahulkhinchi03/checkstyle-test/runs/7958230091?check_suite_focus=true

mvn release:prepare created a commit and push it. This makes it happen itself. And we are committing and pushing it again in the action which make it fail

File: .ci/maven-prepare.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+CURRENT_VERSION=$(xmlstarlet sel -N pom=http://maven.apache.org/POM/4.0.0 \
+                             -t -m pom:project -v pom:version pom.xml | sed "s/-SNAPSHOT//")
+echo CURRENT_VERSION="$CURRENT_VERSION"
+
+if [ "$TARGET_VERSION" != "$CURRENT_VERSION" ]; then
+  echo "[ERROR] Target Version and Current Version doesn't match."
+  exit 1;
+fi
+
+git config --global user.name 'github-actions[bot]'
```
Comment (romani) [inline .ci/maven-prepare.sh:23]: Please make check if such value is not set.

Such shell should be able to be executed by maintainer on his local.
So such command will be damaging for him.

File: .ci/maven-prepare.sh [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+git config --global user.email 'github-actions[bot]@users.noreply.github.com'
+
+SKIP_TEST="-DskipTests -DskipITs"
+SKIP_CHECKSTYLE="-Dcheckstyle.ant.skip=true -Dcheckstyle.skip=true"
+SKIP_OTHERS="-Dpmd.skip=true -Dspotbugs.skip=true -Djacoco.skip=true -Dxml.skip=true"
+
+echo "Version bump in pom.xml (release:prepare) ..."
+mvn -e --no-transfer-progress release:prepare -B -Darguments="$SKIP_TEST $SKIP_CHECKSTYLE \
```
Comment (romani) [inline .ci/maven-prepare.sh:31]: Put `\` after -B.

File: .github/maven_prepare.yml [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+        uses: actions/checkout@v2
+        with:
+          token: ${{ secrets.GITHUB_TOKEN }}
+      - name: Run Shell Script
+        run: |
+          ./.ci/maven-prepare.sh ${{ github.event.inputs.version }}
+      - name: Commit and Push
+        run: |
+              git config --global user.name 'github-actions[bot]'
+              git config --global user.email 'github-actions[bot]@users.noreply.github.com'
```
Comment (romani) [inline .github/maven_prepare.yml:27]: Why twice?
In script and here?

Make such set before script execution, so you will not need to bother about it in script at all

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .github/maven_prepare.yml (hunk 7 above)
Comment (Rahulkhinchi03) [reply .github/maven_prepare.yml]: Changes Done

File: .ci/maven-prepare.sh (hunk 5 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

File: .github/maven_prepare.yml (hunk 7 above)
Comment (Rahulkhinchi03) [reply .github/maven_prepare.yml]: Changes Done

File: .ci/maven-prepare.sh (hunk 6 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

File: .ci/maven-prepare.sh [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+
+if [[ -z $1 ]]; then
+  echo "latest version is not set"
+  exit 1
+fi
+
+TARGET_VERSION=$1
+echo TARGET_VERSION="$TARGET_VERSION"
+
+sudo apt-get install xmlstarlet
```
Comment (romani) [inline .ci/maven-prepare.sh:13]: Move this to action.
Script should be ready to run on local

**Review by romani — CHANGES_REQUESTED** [review]
Items

File: .ci/maven-prepare.sh (hunk 8 above)
Comment (Rahulkhinchi03) [reply .ci/maven-prepare.sh]: Changes Done

**Review by romani — APPROVED** [review]
Ok to merge

**Pull request merged**